network/src/router.rs

```diff
@@ -46,24 +53,86 @@ fn attestation_topic(parent_hash: Hash) -> Hash {
 	BlakeTwo256::hash(&v[..])
 }
 
+fn incoming_message_topic(parent_hash: Hash, parachain: ParaId) -> Hash {
+	let mut v = parent_hash.as_ref().to_vec();
+	parachain.using_encoded(|s| v.extend(s));
+	v.extend(b"incoming");
+
+	BlakeTwo256::hash(&v[..])
+}
+
+/// Receiver for block data.
+pub struct BlockDataReceiver {
+	outer: ::futures::sync::oneshot::Receiver<::futures::sync::oneshot::Receiver<BlockData>>,
+	inner: Option<::futures::sync::oneshot::Receiver<BlockData>>
+}
+
+impl Future for BlockDataReceiver {
+	type Item = BlockData;
+	type Error = io::Error;
+
+	fn poll(&mut self) -> Poll<BlockData, io::Error> {
+		if let Some(ref mut inner) = self.inner {
+			return inner
+				.poll()
+				.map_err(|_| io::Error::new(
+					io::ErrorKind::Other,
+					"Sending end of channel hung up",
+			))
+		}
+		if let Ok(futures::Async::Ready(mut inner)) = self.outer.poll() {
+			let poll_result = inner.poll();
+			self.inner = Some(inner);
+			return poll_result
+				.map_err(|_| io::Error::new(
+					io::ErrorKind::Other,
+					"Sending end of channel hung up",
+			))
+		}
+		Ok(futures::Async::NotReady)
```

Comment:
If `self.outer.poll()` returns `Err` we return `Ok(NotReady)`, which will prevent further polling, right?